File: NiftyOptionCalculator-Android/app/src/main/java/com/niftyoption/calculator/ui/viewmodels/OptionScreenerViewModel.kt
package com.niftyoption.calculator.ui.viewmodels

import androidx.lifecycle.ViewModel
import com.niftyoption.calculator.data.models.MoneynessFilter
import com.niftyoption.calculator.data.models.OptionChainRow
import com.niftyoption.calculator.data.models.OptionData
import com.niftyoption.calculator.data.models.ScreenerFilter
import com.niftyoption.calculator.data.models.ScreenerOptionType
import com.niftyoption.calculator.data.models.ScreenerResult
import com.niftyoption.calculator.data.models.ScreenerSortBy
import com.niftyoption.calculator.domain.BlackScholesEngine
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

class OptionScreenerViewModel : ViewModel() {
    private val _filter = MutableStateFlow(ScreenerFilter())
    val filter: StateFlow<ScreenerFilter> = _filter.asStateFlow()

    private val _results = MutableStateFlow<List<ScreenerResult>>(emptyList())
    val results: StateFlow<List<ScreenerResult>> = _results.asStateFlow()

    private val _resultCount = MutableStateFlow(0)
    val resultCount: StateFlow<Int> = _resultCount.asStateFlow()

    private var optionChain: List<OptionChainRow> = emptyList()
    private var spotPrice: Double = 0.0

    fun setOptionChainData(chain: List<OptionChainRow>, spot: Double) {
        optionChain = chain
        spotPrice = spot
        applyFilters()
    }

    fun updateFilter(newFilter: ScreenerFilter) {
        _filter.value = newFilter
        applyFilters()
    }

    fun resetFilters() {
        _filter.value = ScreenerFilter()
        applyFilters()
    }

    private fun applyFilters() {
        val filter = _filter.value
        val results = mutableListOf<ScreenerResult>()

        for (row in optionChain) {
            // Process call option
            if (filter.optionType != ScreenerOptionType.PUT) {
                row.callOption?.let { option ->
                    if (passesFilters(option, filter)) {
                        results.add(toScreenerResult(option))
                    }
                }
            }
            // Process put option
            if (filter.optionType != ScreenerOptionType.CALL) {
                row.putOption?.let { option ->
                    if (passesFilters(option, filter)) {
                        results.add(toScreenerResult(option))
                    }
                }
            }
        }

        // Sort
        val sorted = when (filter.sortBy) {
            ScreenerSortBy.DELTA -> results.sortedBy { kotlin.math.abs(it.delta) }
            ScreenerSortBy.IV -> results.sortedBy { it.iv }
            ScreenerSortBy.OI -> results.sortedBy { it.oi.toLong() }
            ScreenerSortBy.VOLUME -> results.sortedBy { it.volume.toLong() }
            ScreenerSortBy.PREMIUM -> results.sortedBy { it.ltp }
        }

        _results.value = if (filter.sortAscending) sorted else sorted.reversed()
        _resultCount.value = _results.value.size
    }

    private fun passesFilters(option: OptionData, filter: ScreenerFilter): Boolean {
        // Moneyness filter
        if (filter.moneyness != MoneynessFilter.ALL) {
            val moneyness = option.moneyness
            if (filter.moneyness.displayName != moneyness) return false
        }

        // Delta filter (use absolute value)
        val T = option.timeToExpiryYears
        val absDelta = if (T > 0 && option.impliedVolatility > 0) {
            val greeks = BlackScholesEngine.calculateGreeks(
                option.underlyingValue, option.strikePrice, T, 0.065, option.impliedVolatility, option.optionType
            )
            kotlin.math.abs(greeks.delta)
        } else {
            0.0
        }
        if (absDelta < filter.deltaMin || absDelta > filter.deltaMax) return false

        // IV filter (compare in percentage)
        val ivPct = option.impliedVolatility * 100
        if (ivPct < filter.ivMin || ivPct > filter.ivMax) return false

        // OI filter
        if (option.openInterest < filter.oiMin) return false

        // Volume filter
        if (option.volume < filter.volumeMin) return false

        // Premium filter
        if (option.lastTradedPrice < filter.premiumMin || option.lastTradedPrice > filter.premiumMax) return false

        return true
    }

    private fun toScreenerResult(option: OptionData): ScreenerResult {
        val T = option.timeToExpiryYears
        val greeks = if (T > 0 && option.impliedVolatility > 0) {
            BlackScholesEngine.calculateGreeks(
                option.underlyingValue, option.strikePrice, T, 0.065, option.impliedVolatility, option.optionType
            )
        } else null

        return ScreenerResult(
            strikePrice = option.strikePrice,
            optionType = option.optionType,
            ltp = option.lastTradedPrice,
            iv = option.impliedVolatility,
            delta = greeks?.delta ?: 0.0,
            gamma = greeks?.gamma ?: 0.0,
            theta = greeks?.theta ?: 0.0,
            vega = greeks?.vega ?: 0.0,
            oi = option.openInterest,
            oiChange = option.changeInOI,
            volume = option.volume,
            underlyingValue = option.underlyingValue
        )
    }
}

File: NiftyOptionCalculator-iOS/NiftyOptionCalculator/ViewModels/OptionScreenerViewModel.swift
import Foundation
import Combine

// MARK: - Screener Supporting Types

enum ScreenerOptionType: String, CaseIterable, Identifiable {
    case both = "Both"
    case ce = "CE"
    case pe = "PE"

    var id: String { rawValue }
}

enum MoneynessFilter: String, CaseIterable, Identifiable {
    case all = "All"
    case atm = "ATM"
    case itm = "ITM"
    case otm = "OTM"

    var id: String { rawValue }
}

enum ScreenerSortBy: String, CaseIterable, Identifiable {
    case delta = "Delta"
    case iv = "IV"
    case oi = "OI"
    case volume = "Volume"
    case premium = "Premium"

    var id: String { rawValue }
}

struct ScreenerFilter {
    var optionType: ScreenerOptionType = .both
    var moneyness: MoneynessFilter = .all
    var deltaMin: Double = 0.0
    var deltaMax: Double = 1.0
    var ivMin: Double = 0.0       // percentage (e.g. 0 for 0%)
    var ivMax: Double = 200.0     // percentage (e.g. 200 for 200%)
    var oiMin: Int = 0
    var volumeMin: Int = 0
    var premiumMin: Double = 0.0
    var premiumMax: Double = 100000.0
    var sortBy: ScreenerSortBy = .oi
    var sortAscending: Bool = false
}

struct ScreenerResult: Identifiable {
    let id = UUID()
    let strikePrice: Double
    let optionType: OptionType
    let ltp: Double
    let iv: Double          // decimal (e.g. 0.15)
    let delta: Double
    let gamma: Double
    let theta: Double
    let vega: Double
    let oi: Int
    let oiChange: Int
    let volume: Int
}

// MARK: - OptionScreenerViewModel

@MainActor
final class OptionScreenerViewModel: ObservableObject {

    // MARK: - Published Properties

    @Published var filters = ScreenerFilter()
    @Published var results: [ScreenerResult] = []
    @Published var resultCount: Int = 0

    // MARK: - Private Properties

    private var optionChain: [OptionChainRow] = []
    private var spotPrice: Double = 0

    // MARK: - Public Methods

    /// Loads option chain data and triggers filtering.
    func setData(chain: [OptionChainRow], spot: Double) {
        optionChain = chain
        spotPrice = spot
        applyFilters()
    }

    /// Updates filters and reapplies them.
    func updateFilters(_ newFilters: ScreenerFilter) {
        filters = newFilters
        applyFilters()
    }

    /// Resets all filters to defaults.
    func resetFilters() {
        filters = ScreenerFilter()
        applyFilters()
    }

    // MARK: - Private Methods

    private func applyFilters() {
        var filtered: [ScreenerResult] = []

        for row in optionChain {
            // Process call options
            if filters.optionType == .both || filters.optionType == .ce {
                if let call = row.callOption {
                    if let result = evaluateOption(call) {
                        filtered.append(result)
                    }
                }
            }

            // Process put options
            if filters.optionType == .both || filters.optionType == .pe {
                if let put = row.putOption {
                    if let result = evaluateOption(put) {
                        filtered.append(result)
                    }
                }
            }
        }

        // Sort results
        filtered.sort { a, b in
            let comparison: Bool
            switch filters.sortBy {
            case .delta:
                comparison = abs(a.delta) < abs(b.delta)
            case .iv:
                comparison = a.iv < b.iv
            case .oi:
                comparison = a.oi < b.oi
            case .volume:
                comparison = a.volume < b.volume
            case .premium:
                comparison = a.ltp < b.ltp
            }
            return filters.sortAscending ? comparison : !comparison
        }

        results = filtered
        resultCount = filtered.count
    }

    private func evaluateOption(_ option: OptionData) -> ScreenerResult? {
        // Moneyness filter
        switch filters.moneyness {
        case .all:
            break
        case .atm:
            guard option.isATM else { return nil }
        case .itm:
            guard option.isITM && !option.isATM else { return nil }
        case .otm:
            guard !option.isITM && !option.isATM else { return nil }
        }

        // Compute Greeks via BlackScholesEngine
        let iv = option.impliedVolatility > 0 ? option.impliedVolatility : 0.15
        let T = option.preciseTimeToExpiryYears

        let greeks: GreeksResult
        if T > 0 && iv > 0 {
            switch option.optionType {
            case .call:
                greeks = BlackScholesEngine.calculateCallGreeks(
                    spotPrice: spotPrice,
                    strikePrice: option.strikePrice,
                    timeToExpiry: T,
                    riskFreeRate: 0.065,
                    volatility: iv
                )
            case .put:
                greeks = BlackScholesEngine.calculatePutGreeks(
                    spotPrice: spotPrice,
                    strikePrice: option.strikePrice,
                    timeToExpiry: T,
                    riskFreeRate: 0.065,
                    volatility: iv
                )
            }
        } else {
            greeks = GreeksResult(delta: 0, gamma: 0, theta: 0, vega: 0, rho: 0)
        }

        // Delta filter (use absolute value for comparison)
        let absDelta = abs(greeks.delta)
        guard absDelta >= filters.deltaMin && absDelta <= filters.deltaMax else { return nil }

        // IV filter (compare as percentage)
        let ivPercent = iv * 100.0
        guard ivPercent >= filters.ivMin && ivPercent <= filters.ivMax else { return nil }

        // OI filter
        guard option.openInterest >= filters.oiMin else { return nil }

        // Volume filter
        guard option.volume >= filters.volumeMin else { return nil }

        // Premium filter
        guard option.lastTradedPrice >= filters.premiumMin &&
              option.lastTradedPrice <= filters.premiumMax else { return nil }

        return ScreenerResult(
            strikePrice: option.strikePrice,
            optionType: option.optionType,
            ltp: option.lastTradedPrice,
            iv: iv,
            delta: greeks.delta,
            gamma: greeks.gamma,
            theta: greeks.theta,
            vega: greeks.vega,
            oi: option.openInterest,
            oiChange: option.changeInOI,
            volume: option.volume
        )
    }
}
